File: apps/web/src/lib/types/load.ts
/**
 * Load and LoadStop types for freight shipments
 */

export interface LoadStop {
  id: number;
  stop_id: number;
  sequence_order: number;
  action_type: "pickup" | "delivery" | "both";
  earliest_arrival?: string;
  latest_arrival?: string;
  estimated_dock_hours: number;
  actual_dock_hours?: number;

  // Stop details (joined from stops table)
  stop_name?: string;
  stop_city?: string;
  stop_state?: string;
  stop_address?: string;
}

export interface Load {
  id: number;
  load_id: string;
  load_number: string;
  status: "pending" | "planned" | "active" | "completed" | "cancelled";
  weight_lbs: number;
  commodity_type: "general" | "hazmat" | "refrigerated" | "fragile";
  special_requirements?: string;
  customer_name: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;

  // Relationships
  stops: LoadStop[];
}

export interface LoadListItem {
  id: number;
  load_id: string;
  load_number: string;
  status: string;
  customer_name: string;
  stop_count: number;
  weight_lbs: number;
  commodity_type: string;
}

export interface LoadStopCreate {
  stop_id: string;
  sequence_order: number;
  action_type: "pickup" | "delivery" | "both";
  earliest_arrival?: string;
  latest_arrival?: string;
  estimated_dock_hours: number;
}

export interface LoadCreate {
  load_number: string;
  weight_lbs: number;
  commodity_type: string;
  special_requirements?: string;
  customer_name: string;
  stops: LoadStopCreate[];
}

File: apps/backend/app/models/load.py
"""Load model for freight shipments with pickup/delivery stops."""

from typing import TYPE_CHECKING, List

from sqlalchemy import Float, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base, TimestampMixin

if TYPE_CHECKING:
    from app.models.load_stop import LoadStop
    from app.models.route_plan import RoutePlan


class Load(Base, TimestampMixin):
    """Load model representing a freight shipment with multiple stops."""

    __tablename__ = "loads"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    load_id: Mapped[str] = mapped_column(String(50), unique=True, index=True, nullable=False)

    # Load identity
    load_number: Mapped[str] = mapped_column(String(100), nullable=False)  # Customer reference
    status: Mapped[str] = mapped_column(
        String(50), default="pending", nullable=False, index=True
    )  # pending, planned, active, completed, cancelled

    # Load details
    weight_lbs: Mapped[float] = mapped_column(Float, nullable=False)
    commodity_type: Mapped[str] = mapped_column(
        String(100), nullable=False
    )  # general, hazmat, refrigerated, fragile
    special_requirements: Mapped[str | None] = mapped_column(Text)

    # Customer (simplified for MVP)
    customer_name: Mapped[str] = mapped_column(String(255), nullable=False)

    # Metadata
    is_active: Mapped[bool] = mapped_column(default=True, nullable=False)

    # Relationships
    stops: Mapped[List["LoadStop"]] = relationship(
        "LoadStop",
        back_populates="load",
        cascade="all, delete-orphan",
        order_by="LoadStop.sequence_order",
    )
    route_plans: Mapped[List["RoutePlan"]] = relationship(
        "RoutePlan", back_populates="load", foreign_keys="RoutePlan.load_id"
    )

    def __repr__(self) -> str:
        return (
            f"<Load(id={self.id}, load_id='{self.load_id}', "
            f"load_number='{self.load_number}', status='{self.status}')>"
        )
